src/java/voldemort/store/readonly/swapper/AdminStoreSwapper.java

```diff
@@ -174,14 +168,52 @@ public String call() throws Exception {
-                            storeName,
-                            storeDir,
-                            pushVersion,
-                            timeoutMs);
+
+                    if (attempt > 1) {
+                        logger.info("Fetch attempt " + attempt + "/" + MAX_FETCH_ATTEMPTS + " for " + node.briefToString()
+                                    + ". Will wait " + WAIT_TIME_BETWEEN_FETCH_ATTEMPTS + " ms before going ahead.");
+                        try {
+                            Thread.sleep(WAIT_TIME_BETWEEN_FETCH_ATTEMPTS );
+                        } catch (InterruptedException e) {
+                            throw new VoldemortException(e);
+                        }
+                    }
+
+                    logger.info("Invoking fetch for " + node.briefToString() + " for " + hadoopStoreDirToFetch);
+                    try {
+                        return adminClient.readonlyOps.fetchStore(node.getId(),
+                                                                  storeName,
+                                                                  hadoopStoreDirToFetch,
+                                                                  pushVersion,
+                                                                  timeoutMs);
+                    } catch (AsyncOperationTimeoutException e) {
+                        throw e;
+                    } catch (VoldemortException ve) {
+                        if (attempt >= MAX_FETCH_ATTEMPTS) {
+                            throw ve;
+                        }
+
+                        if (ExceptionUtils.recursiveClassEquals(ve, ExceptionUtils.BNP_SOFT_ERRORS)) {
+                            String logMessage = "Got a " + ve.getClass().getSimpleName() + " from " + node.briefToString()
+                                                + " while trying to fetch store '" + storeName + "'"
+                                                + " (attempt " + attempt + "/" + MAX_FETCH_ATTEMPTS + ").";
+                            AdminClient possiblyRefreshedAdminClient = adminClient;
+                            if (adminClient.isClusterModified()) {
+                                logMessage += " It seems like the cluster.xml state has changed since this AdminClient"
+                                            + " was constructed. Therefore, we will attempt constructing a fresh"
+                                            + " AdminClient and retrying the fetch operation.";
+                                possiblyRefreshedAdminClient = adminClient.getFreshClient();
+                            } else {
+                                logMessage += " The cluster.xml is up to date. We will retry with the same AdminClient.";
+                            }
+                            logger.info(logMessage);
+                            return fetch(hadoopStoreDirToFetch, possiblyRefreshedAdminClient, ++attempt);
```

Comment:
So you have a concern that there could be a stack overflow? Or is it a stylistic preference? I don't think the former is a big issue, given the maximum number of retries?